_posts/2020-10-19-tests-best-practices.md

```diff
@@ -0,0 +1,233 @@
+   changed.
+
+### Assertion:
+Most tests, including their setup, include assertions that check expected
+results from individual operations.
+
+Asserts include convenient checks for various conditions and provide verbose
+information on failures.
+
+### Skipping:
+Skips can be split into two categories: the ones that are hard-coded to skip
+and those that will skip per a specific condition during run-time.
+
+The first fits an expected test to fail scenario, where a test is added to
+show a limitation or a bug in the system, expecting to be fixed in the future.
+
+The second is used in various scenarios, most commonly to detect a platform or
+a missing capability in which the test is not expected to pass, therefore set
+to skip.
+
+An (improved) alternative to the dynamic skip is to classify well the tests
+(e.g. “linux-only”, “feature-foo”) and at test execution, use their
+marking/labeling to filter them out from the run list.
+This leaves the control to the test runner logic and avoids unintentional
+skips to occur, introducing holes in the coverage.
+
+### Focus:
+It is common during development and troubleshooting to run specific tests and
+not the whole suite.
+
+It is expected that tests will correctly operate, running the needed setup and
+teardown associated with them.
+This emphasises that a test should be isolated with no dependencies on other
+tests (as there is no guarantee on how it will be called).
+
+
+## Test Best Practices
+The following best practices are agnostic to languages and frameworks,
+they outline patterns that may be controversial and sometimes not easy to
+apply on some systems.
+
+### Fail First
+A test should fail before it passes otherwise it may end up always passing,
+catching nothing.
+
+### Readability and Structure
+Test code that is easy to read, is easy to understand and maintain.
+
+As with any code base, use well named tests, variables and
+functions. Keep functions small and focused, split into directories and
+files with a good and intuitive structure so it will be easy to track
+tests/helpers and know where to add new ones (or avoid adding duplicate ones).
+
+Note: Avoid generic names like "utils", "common" and alike. They usually start
+small but quickly get filled with unrelated content, becoming a junkyard of
+tests/helpers that do not fit anywhere else.
+
+### Test body and Fixture distinction
+Place the operations into the right step (setup, test body, teardown).
+
+Indirect operations that are not in the main focus of the test, better be
+placed in the setup or teardown fixtures and not in the test body.
+
+As an example, testing if an entity can be successfully created fits a test
+body.
+However, testing if an entity can communicate with another is about
+the communication check and not its creation or deletion, therefore, the
+creation will better fit the setup (and deletion into the teardown) while the
+communication check fits the test body.
+
```

Comment:
moving your verbal examle to pseudo code example
example (1): entity creation test
```
Test setup:

Test body:
create_entity()
check_entity_creation_success()
```

example (2): entity communication test
```
Test setup:
create_entity()
check_entity_creation_success()

Test body:
check_entity_connectivity()
```

Reply:
Got it, thanks!